File: chazm-model/src/main/java/runtimemodels/chazm/model/event/SpecificationGoalEvent.java
package runtimemodels.chazm.model.event;

import com.google.inject.assistedinject.Assisted;
import runtimemodels.chazm.api.entity.SpecificationGoal;
import runtimemodels.chazm.api.id.UniqueId;
import runtimemodels.chazm.message.M;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;
import java.util.Objects;

/**
 * The {@linkplain SpecificationGoalEvent} class indicates that there is an update about an {@linkplain SpecificationGoal}.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class SpecificationGoalEvent extends AbstractEvent {

    private static final long serialVersionUID = 602095857291387756L;
    private final UniqueId<SpecificationGoal> id;
    private transient Integer hashCode = null;
    private transient String toString = null;

    @Inject
    SpecificationGoalEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final SpecificationGoal goal) {
        super(category);
        id = goal.getId();
    }

    /**
     * Returns a {@linkplain UniqueId} that represents a {@linkplain SpecificationGoal}.
     *
     * @return a {@linkplain UniqueId}.
     */
    public UniqueId<SpecificationGoal> getId() {
        return id;
    }

    @Override
    public boolean equals(final Object object) {
        if (object instanceof SpecificationGoalEvent) {
            final SpecificationGoalEvent event = (SpecificationGoalEvent) object;
            return super.equals(event) && getId().equals(event.getId());
        }
        return false;
    }

    @Override
    public int hashCode() {
        if (hashCode == null) {
            hashCode = Objects.hash(getCategory(), getId());
        }
        return hashCode;
    }

    @Override
    public String toString() {
        if (toString == null) {
            toString = M.EVENT_WITH_1_ID.get(super.toString(), getId());
        }
        return toString;
    }

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/SpecificationGoalEvent.kt
package runtimemodels.chazm.model.entity

import com.google.inject.assistedinject.Assisted
import runtimemodels.chazm.api.entity.SpecificationGoal
import runtimemodels.chazm.api.entity.SpecificationGoalId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [SpecificationGoalEvent] class indicates that there is an update about an [SpecificationGoal].
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class SpecificationGoalEvent @Inject internal constructor(
    @Assisted category: EventType,
    @Assisted goal: SpecificationGoal
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [SpecificationGoal].
     *
     * @return a [UniqueId].
     */
    val id: SpecificationGoalId = goal.id

    override fun equals(other: Any?): Boolean {
        if (other is SpecificationGoalEvent) {
            return super.equals(other) && id == other.id
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, id)

    override fun toString(): String = M.EVENT_WITH_1_ID[super.toString(), id]

    companion object {
        private const val serialVersionUID = 602095857291387756L
    }

}
